src/main/java/io/pinecone/PineconeDataPlaneClient.java

```diff
@@ -0,0 +1,222 @@
+import java.util.List;
+
+public class PineconeDataPlaneClient {
+    private VectorServiceGrpc.VectorServiceBlockingStub blockingStub;
+
+    public PineconeDataPlaneClient(VectorServiceGrpc.VectorServiceBlockingStub blockingStub) {
+        if (blockingStub == null) {
+            throw new PineconeValidationException("BlockingStub cannot be null.");
+        }
+
+        this.blockingStub = blockingStub;
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values) {
+        return upsert(id, values, null, null, null, null);
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values,
+                                 String namespace) {
+        return upsert(id, values, null, null, null, namespace);
+    }
+
+    public UpsertResponse upsert(String id,
+                                 List<Float> values,
+                                 List<Long> sparseIndices,
+                                 List<Float> sparseValues,
+                                 com.google.protobuf.Struct metadata,
+                                 String namespace) {
+        UpsertRequest.Builder upsertRequest = UpsertRequest.newBuilder();
+
+        if (id == null || id.isEmpty() || values == null || values.isEmpty()) {
+            throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both id and values are provided.");
+        }
+
+        Vector.Builder vectorBuilder = Vector.newBuilder()
+                .setId(id)
+                .addAllValues(values);
+
+        if (sparseIndices != null && sparseValues != null && sparseIndices.size() != sparseValues.size()) {
+            throw new PineconeValidationException("Invalid Upsert Request. Please ensure that both sparse indices and values are of the same length.");
+        }
+
+        if (sparseIndices != null && !sparseIndices.isEmpty()) {
+            vectorBuilder.setSparseValues(SparseValues.newBuilder()
+                    .addAllIndices(convertUnsigned32IntToSigned32Int(sparseIndices))
+                    .addAllValues(sparseValues)
+                    .build());
+        }
+
+        if (metadata != null) {
+            vectorBuilder.setMetadata(metadata);
+        }
+
+        upsertRequest.addVectors(vectorBuilder.build());
+
+        if (namespace != null && !namespace.isEmpty()) {
+            upsertRequest.setNamespace(namespace);
+        }
+
+        return blockingStub.upsert(upsertRequest.build());
+    }
+
+    public QueryResponse query(String id,
+                               String namespace,
+                               int topK,
+                               Struct filter,
+                               boolean includeValues,
+                               boolean includeMetadata,
+                               List<Float> vectors,
+                               List<Long> sparseIndices,
+                               List<Float> sparseValues) {
```

Comment:
I would make some changes here:

- Put topK first, since it is always required
- Change `vectors` to `vector`. This aligns with the [python arg naming.](https://github.com/pinecone-io/pinecone-python-client/blob/main/pinecone/data/index.py#L330)
- Prioritize `values` over `id` because I think that is the more commonly used way of querying.
- Make `includeValues` and `includeMetadata` nullable, and implement defaults of `false` inside this method. This is less important for the others, but I think `null` is a nice shorthand for `I don't care, use a default value` that we can easily implement.

These changes will make common usage look like `query(10, Arrays.asList(0.1f, 0.2f, 0.3f), null, null, null, null, null, null, null)` grouping all nulls toward the end.

```suggestion
    public QueryResponse query(
                               int topK,
                               List<Float> vector,
                               List<Long> sparseIndices,
                               List<Float> sparseValues
                               String id,
                               String namespace,
                               Struct filter,
                               Boolean includeValues,
                               Boolean includeMetadata,
) {
            // todo implementation
    }
 ```


Then I think this would be better if we added some helpers such as `queryByVectorId` and `queryByVectorValues` methods. I think this makes the process of adding overloaded methods much easier, since id & vector you can ignore some fields that aren't relevant to that scenario.

```java
    public QueryResponse queryByVectorId(
                               int topK,
                               String id,
                               String namespace,
                               Struct filter,
                               Boolean includeValues,
                               Boolean includeMetadata,) {
            // todo implementation
    }

    public QueryResponse queryByVectorId(
                               int topK,
                               String id,
                               String namespace,
                               Struct filter
    ) {
            // todo implementation where includeValues, includeMetadata are null
    }

    public QueryResponse queryByVectorId(
                               int topK,
                               String id,
                               String namespace,
    ) {
            // todo implementation where includeValues, includeMetadata, filter are null
    }

    public QueryResponse queryByVectorId(
                               int topK,
                               String id
    ) {
            // todo implementation where includeValues, includeMetadata, filter, namespace are null
    }
```

Reply:
This is really great feedback, thanks so much!!